File: datacenter/tests/datapoint-source.test.ts
import { describe, expect, test } from 'vitest'
import {
  resolveDatapointSourceDisplayName,
  resolveDatapointSourceNavigation,
} from '../src/models/datapoint-source'

describe('datapoint source navigation', () => {
  test('计算输出定位到计算单元', () => {
    expect(
      resolveDatapointSourceNavigation({
        sourceType: 'calc.output',
        sourceId: 'compute-1',
      }),
    ).toEqual({
      module: 'compute',
      objectId: 'compute-1',
      label: '计算单元',
      actionLabel: '打开计算单元',
    })
  })

  test('工业采集点定位到采集连接', () => {
    expect(
      resolveDatapointSourceNavigation({
        sourceType: 'collector.point',
        accessSourceId: 'collector-1',
      }),
    ).toMatchObject({ module: 'industrial-collector', objectId: 'collector-1' })
  })

  test.each(['db.query', 'mqtt.subscription', 'mqtt.tag'])(
    '%s 使用后端解析的接入源连接',
    (sourceType) => {
      expect(
        resolveDatapointSourceNavigation({ sourceType, accessSourceId: 'connection-1' }),
      ).toMatchObject({ module: 'access-source', objectId: 'connection-1' })
    },
  )

  test.each(['http.request', 'websocket.session', 'realtime.key', 'kafka.field', 'kafka.raw'])(
    '%s 可从 sourceId 回退解析连接',
    (sourceType) => {
      expect(
        resolveDatapointSourceNavigation({ sourceType, sourceId: 'connection-1' }),
      ).toMatchObject({ module: 'access-source', objectId: 'connection-1' })
    },
  )

  test('无可导航来源时不产生错误目标', () => {
    expect(resolveDatapointSourceNavigation({ sourceType: 'static.var' })).toBeNull()
    expect(resolveDatapointSourceDisplayName({ sourceType: 'static.var' })).toBe('静态变量')
  })
})

File: data_service/internal/service/datapoint_source_test.go
package service

import (
	"context"
	"testing"

	"github.com/indu-forge/data_service/internal/repository"
)

func TestToDataPointListItemKeepsSourceObjectID(t *testing.T) {
	sourceID := "00000000-0000-0000-0000-000000000001"
	service := &DataPointService{}
	item := service.toDataPointListItem(context.Background(), "project-1", repository.DataPointRecord{
		SourceType: "calc.output",
		SourceID:   &sourceID,
	})

	if item.SourceID == nil || *item.SourceID != sourceID {
		t.Fatalf("sourceId = %#v, want %q", item.SourceID, sourceID)
	}
}

func TestSourceTypeUsesSourceIDAsConnectionIncludesProtocolOutputs(t *testing.T) {
	for _, sourceType := range []string{"http.request", "websocket.session", "realtime.key", "kafka.field", "kafka.raw"} {
		if !sourceTypeUsesSourceIDAsConnection(sourceType) {
			t.Fatalf("source type %q should use sourceId as connection id", sourceType)
		}
	}

	for _, sourceType := range []string{"calc.output", "db.query", "mqtt.tag", "collector.point"} {
		if sourceTypeUsesSourceIDAsConnection(sourceType) {
			t.Fatalf("source type %q must not use sourceId as connection id", sourceType)
		}
	}
}

func TestDataPointQueryParametersAcceptsSavedDefinitions(t *testing.T) {
	parameters, err := dataPointQueryParameters(map[string]any{
		"parameters": []any{
			map[string]any{"name": "withoutDefault"},
			map[string]any{"name": "limit", "default": 10},
		},
	})
	if err != nil {
		t.Fatalf("dataPointQueryParameters returned error: %v", err)
	}
	if len(parameters) != 1 || parameters["limit"] != 10 {
		t.Fatalf("parameters = %#v, want only explicit default", parameters)
	}

	empty, err := dataPointQueryParameters(map[string]any{"parameters": []any{}})
	if err != nil || len(empty) != 0 {
		t.Fatalf("empty definitions = %#v, %v; want empty map", empty, err)
	}
}

func TestQueryDatasetValueKeepsOnlyReturnedRowCount(t *testing.T) {
	dataset := queryDatasetValue(&QueryExecutionResult{
		Data:        []map[string]any{{"id": 1}},
		Columns:     []string{"id"},
		RowCount:    1,
		Truncated:   true,
		TruncatedBy: "rows",
	})
	if dataset["rowCount"] != 1 {
		t.Fatalf("dataset = %#v, want returned row count", dataset)
	}
	for _, omitted := range []string{"total", "truncated"} {
		if _, exists := dataset[omitted]; exists {
			t.Fatalf("dataset must omit %s: %#v", omitted, dataset)
		}
	}
}
